Sources/PitchSpeller/Data Structures/FlowNetwork.swift

```diff
@@ -32,18 +39,78 @@ public struct FlowNetwork <Value: Hashable>: Hashable {
     /// reached zero in the flow-propagation process.
     ///
     /// - TODO: Add backflow to reversed edges.
-    public var residualNetwork: Graph<Value> {
+    public var residualNetwork: Graph<Node> {
         var residualNetwork = graph
         while let path = residualNetwork.shortestPath(from: source, to: sink) {
             residualNetwork.insertPath(path.map { $0 - maximumFlow(of: path) })
         }
         return residualNetwork
     }
+    
+    /// - Returns: (0) The maximum flow of the network and (1) the residual network produced after
+    /// pushing all possible flow from source to sink (while satisfying flow constraints) - with
+    /// saturated edges flipped and all weights removed.
+    var solvedForMaximumFlow: (Double, UnweightedGraph<Node>) {
+        var totalFlow = 0.0
+        var residualNetwork = directedGraph
+        
+        func findAugmentingPath () -> Bool {
+            guard let path = residualNetwork.shortestUnweightedPath(from: source, to: sink) else {
+                return false
+            }
+            pushFlow(through: path)
+            return true
+        }
+            
+        func pushFlow (through path: UnweightedGraph<Node>.Path) {
+            let minimumEdge = (path.adjacents.compactMap {
+                residualNetwork.weight($0)
+                }.min())!
+            totalFlow += minimumEdge
+            path.adjacents.forEach {
+                residualNetwork.updateEdge($0, with: {
+                    minuend in
+                    minuend - minimumEdge
+                })
+            }
+        }
+        
+        func addBackEdges () {
+            residualNetwork.edges.map {
+                $0.nodes
+                }.filter {
+                    let weight = residualNetwork.weight($0)!
+                    return weight < 0.001 && weight > -0.001
+                }.forEach {
+                    residualNetwork.flipEdge(at: $0)
+            }
+        }
+        
+        while findAugmentingPath() { continue }
+        addBackEdges()
+        return (totalFlow, DirectedGraph<Node>.unWeightedVersion(of: residualNetwork))
+    }
 
     /// - Returns: The two partitions on either side of the s-t cut.
-    public var partitions: (source: Graph<Value>, sink: Graph<Value>) {
+    public var partitions: (source: Graph<Node>, sink: Graph<Node>) {
         return (graph(sourceReachableNodes), graph(sinkReachableNodes.reversed()))
     }
+    
+    /// - Returns: A minimum cut with nodes included on the `sink` side in case of a
+    /// tiebreak (in- and out- edges saturated).
+    public var minimumCut: (Set<Node>, Set<Node>) {
+        return (sourceSideNodes, notSourceSideNodes)
+    }
+    
+    /// - Returns: Nodes in residual network reachable from the `source`
+    private var sourceSideNodes: Set<Node> {
+        return Set(solvedForMaximumFlow.1.breadthFirstSearch(from: source))
```

Comment:
`solvedForMaximumFlow` could return a labeled tuple, which will make this a little less cryptic.

Reply:
👍